cookbook/case_studies/ml_training/spark_horovod/keras_spark_rossmann_estimator.py

```diff
@@ -36,421 +38,608 @@
+    vocab = build_vocabulary(
+        train_df.select(*CATEGORICAL_COLS)
+        .unionAll(test_df.select(*CATEGORICAL_COLS))
+        .cache(),
+    )
 
     # Cast continuous columns to float & lookup categorical columns.
-    train_df = cast_columns(train_df, continuous_cols + ['Sales'])
+    train_df = cast_columns(train_df, CONTINUOUS_COLS + ["Sales"])
     train_df = lookup_columns(train_df, vocab)
-    test_df = cast_columns(test_df, continuous_cols)
+    test_df = cast_columns(test_df, CONTINUOUS_COLS)
     test_df = lookup_columns(test_df, vocab)
 
     # Split into training & validation.
     # Test set is in 2015, use the same period in 2014 from the training set as a validation set.
     test_min_date = test_df.agg(F.min(test_df.Date)).collect()[0][0]
     test_max_date = test_df.agg(F.max(test_df.Date)).collect()[0][0]
     one_year = datetime.timedelta(365)
-    train_df = train_df.withColumn('Validation',
-                                   (train_df.Date > test_min_date - one_year) & (train_df.Date <= test_max_date - one_year))
+    train_df = train_df.withColumn(
+        "Validation",
+        (train_df.Date > test_min_date - one_year)
+        & (train_df.Date <= test_max_date - one_year),
+    )
 
     # Determine max Sales number.
     max_sales = train_df.agg(F.max(train_df.Sales)).collect()[0][0]
 
     # Convert Sales to log domain
-    train_df = train_df.withColumn('Sales', F.log(train_df.Sales))
+    train_df = train_df.withColumn("Sales", F.log(train_df.Sales))
 
-    print('===================================')
-    print('Data frame with transformed columns')
-    print('===================================')
+    print("===================================")
+    print("Data frame with transformed columns")
+    print("===================================")
     train_df.show()
 
-    print('================')
-    print('Data frame sizes')
-    print('================')
+    print("================")
+    print("Data frame sizes")
+    print("================")
+
     train_rows = train_df.filter(~train_df.Validation).count()
     val_rows = train_df.filter(train_df.Validation).count()
     test_rows = test_df.count()
-    print('Training: %d' % train_rows)
-    print('Validation: %d' % val_rows)
-    print('Test: %d' % test_rows)
 
-    # ============== #
-    # MODEL TRAINING #
-    # ============== #
+    print("Training: %d" % train_rows)
+    print("Validation: %d" % val_rows)
+    print("Test: %d" % test_rows)
+
+    return max_sales, vocab, train_df, test_df
 
-    print('==============')
-    print('Model training')
-    print('==============')
 
+def train(
+    max_sales: float,
+    vocab: Dict[str, List[Any]],
+    hp: Hyperparameters,
+    work_dir: FlyteDirectory,
+    train_df: pyspark.sql.DataFrame,
+    working_dir: FlyteDirectory,
+):
+    print("==============")
+    print("Model training")
+    print("==============")
 
     def exp_rmspe(y_true, y_pred):
```

Comment:
can we just make this function at the module level? Why make them to be function local?

Reply:
I can place `exp_rmspe` function at the module level. However, the problem is with the other custom object function: `act_sigmoid_scaled`. 

```python
def act_sigmoid_scaled(x):
    """Sigmoid scaled to logarithm of maximum sales scaled by 20%."""
    return tf.nn.sigmoid(x) * tf.math.log(max_sales) * 1.2
```

The latter function cannot be placed at the module level because we cannot explicitly send `max_sales` as an argument. This is because we use `act_sigmoid_scaled` as an activation function (note the usage of function name rather than function call in `output = Dense(1, activation=act_sigmoid_scaled)(x)`).

That's why I placed `exp_rmspe` and `act_sigmoid_scaled` within `train`; the two of them are custom objects and putting one at the module level, and the other within the function doesn't really add up.

Reply:
i think we should doc this?